File: Problems/394_Decode_String.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        for c in s:
            if c!=']':
                stack.append(c)
            else:
                string = ""
                ch = stack.pop()
                while ch.isalpha():
                    string = ch+string
                    ch = stack.pop()
                num = ""
                ch = stack.pop()
                while ch.isnumeric():
                    num = ch+num
                    if stack and stack[-1].isnumeric():
                        ch = stack.pop()
                    else:
                        break
                num = int(num)
                stack.append(string*num)
        return ''.join(stack)

if __name__ == '__main__':
    sol = Solution()
    s = "3[a]2[bc]"
    print(sol.decodeString(s))

File: Problems/394_Decode_String.cpp
#include<string>
#include<stack>
#include<iostream>

using namespace std;
class Solution {
public:
    string decodeString(string s) {
        stack<char> stck;
        for(char c: s){
            if(c!=']') stck.push(c);
            else{
                string str = "";
                while(stck.top()!='['){
                    str = stck.top()+str;
                    stck.pop();
                }
                stck.pop();
                string num = "";
                while(!stck.empty() && isdigit(stck.top())){
                    num = stck.top()+num;
                    stck.pop();
                }
                for(int i=0; i<stoi(num); i++){
                    for(int j=0; j<str.length(); j++){
                        stck.push(str[j]);
                    }
                }
            }
        }
        string ans = "";
        while(!stck.empty()){
            ans = stck.top()+ans;
            stck.pop();
        }
        return ans;
    }
};

int main(){
    string s = "3[a]2[bc]";
    Solution sol;
    string ans = sol.decodeString(s);
    cout<<ans<<endl;
    return 0;
}